### PR TITLE
[Form] Add input  + regions options to TimezoneType

### PR DESCRIPTION
**Pull request by ro0NL**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | #... 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/8223

I want to use `\DateTimeZone` as a model format, with only european timezones in the form available. Hence the new options.

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 1]
```diff
@@ -27,6 +30,41 @@ class TimezoneType extends AbstractType implements ChoiceLoaderInterface
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $this->regions = $options['regions'];
+
+        if ('datetimezone' === $options['input']) {
+            $builder->addModelTransformer(new CallbackTransformer(function ($timezone) {
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: I think to create an individual class for this model transformer it's the common implementation (i.e. into core) and both will be easier to maintain. `CallbackTransformer` it's commonly used in tests to simplify.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 2]
```diff
@@ -72,7 +116,7 @@ public function loadChoiceList($value = null)
             return $this->choiceList;
         }
 
-        return $this->choiceList = new ArrayChoiceList($this->getTimezones(), $value);
+        return $this->choiceList = new ArrayChoiceList($this->getTimezones($this->regions), $value);
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: This is an issue. The choice list is cached and reused. By doing this, you'll always get the same list when using the type multiple times in a form or nested forms.

@HeahDude could probably give you more inputs than me for about this right now, because I think it was something he had to consider when working on https://github.com/symfony/symfony/pull/18334.
You should also avoid storing options in type properties, as form type are instantiated once, it acts like a global state for the type.

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Right :) if we can solve it simply thats nice, otherwise lets leave it out. I can validate it additionally as well, yet it's nice to have.

Btw shouldnt this be `self::getTimezones()`?

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: > Btw shouldnt this be self::getTimezones()?

Right :)

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: See #23652

Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: > You should also avoid storing options in the form type's properties, as form types are instantiated once, it acts like a global state for the type and all its usages.

An option could be return a new default choice loader instance instead of the current one `return $this;`? e.g.:
```php
return new CallbackChoiceLoader(function () use ($options) {
    return self::getTimezones($options['region']);
});
```
But also, it involves remove the actual `ChoiceLoaderInterface` implementation and maybe I'm missing something...

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: @yceruto is right, we now need a dynamic `TimezonesChoiceLoader` implementation taking the option in its constructor.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php (hunk 1 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: 👍

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 3]
```diff
@@ -27,6 +30,41 @@ class TimezoneType extends AbstractType implements ChoiceLoaderInterface
...
      */
     private $choiceList;
+    private $regions;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $this->regions = $options['regions'];
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Should be removed.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 4]
```diff
@@ -27,6 +30,41 @@ class TimezoneType extends AbstractType implements ChoiceLoaderInterface
...
+                    return;
+                }
+                if (!is_string($value)) {
+                    throw new TransformationFailedException('Expected a string.');
+                }
+
+                try {
+                    return new \DateTimeZone($value);
+                } catch (\Exception $e) {
+                    throw new TransformationFailedException('Expected a known timezone identifier');
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Missing dot.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php (hunk 1 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Done

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php (hunk 2 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: @HeahDude what about the ChoiceLoaderInterface implem. Can we simply drop it?

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Just moved I'd say ;).

**Comment by ro0NL** [comment]
Not sure, but we might consider favoring datetimezone as default input format; so it's consistent with datetime type.

Also i need the next step basically; integrate timezone type with datetime. A `with_timezone` option or so..  that would save me a few steps :)

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\Extension\Core\DataTransformer;
+
+use Symfony\Component\Form\DataTransformerInterface;
+use Symfony\Component\Form\Exception\TransformationFailedException;
+
+/**
+ * Transforms between a date string and a DateTimeZone object.
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php]: "between a timezone identifier string and a DateTimeZone object"

**Comment by ro0NL** [comment]
Updated so that `ChoiceLoaderInterface` implem is deprecated and uses a CallbackLoader directly. Let me know if this is OK.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 6]
```diff
@@ -41,10 +56,18 @@ public function configureOptions(OptionsResolver $resolver)
                     return null;
                 }
 
-                return $this;
+                return new CallbackChoiceLoader(function () use ($options) {
+                    return self::getTimezones($options['regions']);
+                });
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: I would add `$regions = $options['regions']` and pass that only variable to the `use` statement.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 7]
```diff
@@ -139,6 +170,6 @@ private static function getTimezones()
             $timezones[$region][str_replace('_', ' ', $name)] = $timezone;
         }
 
-        return $timezones;
+        return 1 === count($timezones) ? reset($timezones) : $timezones;
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: `1 === count($timezones) ? $timezones : reset($timezones)`?

**Review by HeahDude — COMMENTED** [review]
Looks good!

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php (hunk 2 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: It basically has not purpose anymore.. not sure we should preserve its api in any way (i.e. forward to the callback loader or so).

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php (hunk 7 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: That would be weird.. :P

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Ah yes sorry about that.

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: But we need an array all the time right? Why this change?

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: We get a flat array (1 region) or a nested one (multiple regions). Let me explain with a picture :)

image

It's just cosmetic sugar. IMHO.

Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Alright! Thanks for clarifying.

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 8]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+3.4.0
+-----
+
+ * deprecated `ChoiceLoaderInterface` implementation in `TimezoneType`
```
Comment (HeahDude) [inline src/Symfony/Component/Form/CHANGELOG.md]: Missing dot.

File: UPGRADE-3.4.md [hunk 9]
```diff
@@ -18,6 +18,11 @@ Finder
    deprecated and will be removed in 4.0 as it used to fix a bug which existed
    before version 5.5.23/5.6.7.
 
+Form
+----
+
+  * Deprecated `ChoiceLoaderInterface` implementation in `TimezoneType`.
```
Comment (HeahDude) [inline UPGRADE-3.4.md]: You should also document the new options here.

File: src/Symfony/Component/Form/CHANGELOG.md (hunk 8 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/CHANGELOG.md]: Well.. it's consistent with the current format :)

Shall add the new options here as well :+1:

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 10]
```diff
@@ -41,10 +56,20 @@ public function configureOptions(OptionsResolver $resolver)
...
 
-                return $this;
+                $regions = $options['regions'];
+
+                return new CallbackChoiceLoader(function () use ($regions) {
+                    return self::getTimezones($regions);
+                });
             },
             'choice_translation_domain' => false,
+            'input' => 'string',
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Something else is troubling me. With this new implementation can we still use this choice field as multiple? I guess we should handle an array of timezones too to be fully BC.

File: UPGRADE-3.4.md (hunk 9 above)
Comment (ro0NL) [inline UPGRADE-3.4.md]: New options are just new features, so not required for any upgrade step. I tend to doc them in Form/CHANGELOG.

File: src/Symfony/Component/Form/CHANGELOG.md (hunk 8 above)
Comment (HeahDude) [inline src/Symfony/Component/Form/CHANGELOG.md]: right and not where I commented.

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php (hunk 10 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: Good catch. Will look into that :+1:

**Review by HeahDude — APPROVED** [review]

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+        }
+
+        if (!is_string($value)) {
+            throw new TransformationFailedException('Expected a string.');
+        }
+
+        try {
+            return new \DateTimeZone($value);
+        } catch (\Exception $e) {
+            throw new TransformationFailedException('Expected a known timezone identifier.');
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php]:
```diff
- throw new TransformationFailedException('Expected a known timezone identifier.');
+ throw new TransformationFailedException('Expected a known timezone identifier.', $e->getCode(), $e);
```

(same for every exception caught IMHO. Eventually, you don't even need to set a custom message, but just re-use `$e->getMessage()` 😄 . See [`DateIntervalToArrayTransformer`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToArrayTransformer.php#L169) for instance)

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+            return;
+        }
+
+        if ($this->multiple) {
+            if (!is_array($value)) {
+                throw new TransformationFailedException('Expected an array.');
+            }
+
+            return array_map(function ($value) {
+                if (!is_string($value)) {
```
Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php]: You could avoid duplicating this logic below by adding using a private method... or either by reusing a `DateTimeZoneToStringTransformer` without the `$multiple` flag.

**Review by ogizanagi — COMMENTED** [review]
It would be great to also test the transformer separately :)

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php (hunk 12 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php]: Yeah.. wasnt sure :P resuing `new self` sounds nice.

File: UPGRADE-4.0.md [hunk 13]
```diff
@@ -228,6 +228,8 @@ Form
    ));
    ```
 
+ * Removed `ChoiceLoaderInterface` implementation in `TimezoneType`.
```
Comment (ogizanagi) [inline UPGRADE-4.0.md]: I'm not sure we need this in the `UPGRADE.md` file. Unless there is a new code path to show.

**Review by ogizanagi — APPROVED** [review]
Nice :)

Failure unrelated.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+            return;
+        }
+
+        if ($this->multiple) {
+            if (!is_array($dateTimeZone)) {
+                throw new TransformationFailedException('Expected an array.');
+            }
+
+            return array_map(function ($dateTimeZone) {
+                return (new self())->transform($dateTimeZone);
```
Comment (HeahDude) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php]: An object is constructed at each iteration :/. What about:
```php
$transformer = new self();

return array_map(function ($dateTimeZone) use ($transformer) {
    return $transformer->transform($dateTimeZone);
// ...
```

File: UPGRADE-4.0.md (hunk 13 above)
Comment (HeahDude) [inline UPGRADE-4.0.md]: The upgrade path is:
```
Before:
class MyTimezoneType extends TimezoneType
{
    public function loadChoices()
    {
        // override the method
    }
}

After:
class MyTimezoneType extends AbstractType
{
    public function. getParent()
    {
        return TimezoneType::class;
    }

    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefault('choice_loader', ...); // override the option instead)
    }
}
```

Comment (ro0NL) [inline UPGRADE-4.0.md]: `Removed `ChoiceLoaderInterface` implementation in `TimezoneType`. Use the "choice_loader" option instead.`

?

Comment (ogizanagi) [inline UPGRADE-4.0.md]: I like both. Let's do both? 😄 

* Removed ``ChoiceLoaderInterface`` implementation in ``TimezoneType``. Use the "choice_loader" option instead:

  Before:
  
  ```php
  class MyTimezoneType extends TimezoneType
  {
      public function loadChoices()
      {
          // override the method
      }
  }
  ```
  
  After:
  
  ```php
  
  class MyTimezoneType extends AbstractType
  {
      public function. getParent()
      {
          return TimezoneType::class;
      }
  
      public function configureOptions(OptionsResolver $resolver)
      {
          $resolver->setDefault('choice_loader', ...); // override the option instead)
      }
  }
  ```

Comment (yceruto) [inline UPGRADE-4.0.md]: +1 for Before/After code also ;)

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        if ($this->multiple) {
+            if (!is_array($dateTimeZone)) {
+                throw new TransformationFailedException('Expected an array.');
+            }
+
+            $transformer = new self();
+
+            return array_map(function ($dateTimeZone) use ($transformer) {
+                return $transformer->transform($dateTimeZone);
+            }, $dateTimeZone);
```
Comment (yceruto) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php]: minor simplification `return array_map(array($transformer, 'transform'), $dateTimeZone);`? however, the actual is fine to me ;)

**Review by yceruto — APPROVED** [review]

File: UPGRADE-4.0.md (hunk 13 above)
Comment (ro0NL) [inline UPGRADE-4.0.md]: After would be

```
class MyTimezoneType extends AbstractType implements ChoiceLoaderInterface
// ...
$resolver->setDefault('choice_loader', $this);
```
right?

Comment (ogizanagi) [inline UPGRADE-4.0.md]: Not necessarily. I think @HeahDude 's suggestion is enough to understand.

Comment (ro0NL) [inline UPGRADE-4.0.md]: btw, while at it, is it worth to do this deprecation for other types? LanguageType and co. could do the same approach =/

It seems a way simpler implem. (CallbackChoiceLoader) compared to to ArrayChoiceLoader.

**Comment by nicolas-grekas** [comment]
small conflict to fix

File: src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php [hunk 16]
```diff
@@ -25,12 +28,24 @@ class TimezoneType extends AbstractType implements ChoiceLoaderInterface
...
+     * @deprecated since version 3.4, to be removed in 4.0
      */
     private $choiceList;
 
     /**
      * {@inheritdoc}
      */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        if ('datetimezone' === $options['input']) {
```
Comment (Tobion) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: How about we name the option value `DateTimeZone` so people can use `['input' => \DateTimeZone::class]`? This avoids typos and makes things more explicit as the value directly maps to a class.

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: I like this suggestion and we could also apply it for other `Date*Type` having this input option.
But I'd keep `datetimezone` as alias for consistency, as I'm not convinced deprecating old option values for other types would be worth it and such aliases are handful too (for configuration based form generation like in EasyAdmin for instance).

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: so.. strtolower? :)

Comment (ogizanagi) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: strtolowering the whole FQCN won't help you much ;)

Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/Type/TimezoneType.php]: i meant `if ('datetimezone' === strtolower($options['input'])) {`..

then again im not sure using class-ish values for `input` makes sense (compared to `data_class`). It kinda implies it supports class names, which it doesnt.

File: src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php (hunk 15 above)
Comment (ro0NL) [inline src/Symfony/Component/Form/Extension/Core/DataTransformer/DateTimeZoneToStringTransformer.php]: So update? Kinda like the one-liner :)

**Comment by ro0NL** [comment]
Should be good :)

**Comment by ogizanagi** [comment]
Thanks @ro0NL.
